File: 092. Reverse Linked List II/Python/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        fake_head = ListNode(0)
        fake_head.next = head
        start = fake_head
        n -= m
        m -= 1
        while m:
            start = start.next
            m -= 1
        current = start.next
        while n:
            temp = start.next
            start.next = current.next
            current.next = current.next.next
            start.next.next = temp
            n -= 1
        return fake_head.next

File: 092. Reverse Linked List II/Java/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode fakeHead = new ListNode(0);
        fakeHead.next = head;
        ListNode start = fakeHead;
        n -= m;
        m--;
        while (m != 0) {
            start = start.next;
            m--;
        }
        ListNode current = start.next;
        while (n != 0) {
            ListNode temp = start.next;
            start.next = current.next;
            current.next = current.next.next;
            start.next.next = temp;
            n--;
        }
        return fakeHead.next;
    }
}